**Setup by Importing modules**

In [7]:
import pandas as pd
import numpy as np
import janitor

**Reading raw dirty data and storing it into Data folder on my repo**

In [ ]:


#creating a variable source_file and assigning it file path of 
# data_cleaning_challenge.xlsx

source_file = "C:/Users/njagi/Downloads/data_cleaning_challenge.xlsx"  #had to change '\' to '/' to assign the file path


#reading the dataset while specifing the sheet i need only

df = pd.read_excel(source_file, sheet_name = 'pharma_data')

df.head()

,Trial ID,Drug Name,Company,Trial Phase,Result,Start Date,End Date,Participants,Dosage (mg),Location
0,T1000,Metformin,Pfizer,Phase II,Ongoing,30/09/2022,2023-04-13,4084,376.55,Nairobi
1,T1001,Paracetamol,Johnson & Johnson,Phase III,SUCESS,01/02/2024,13/07/2024,3301,414.65,Kisumu
2,T1002,Aspirin,Novartis,Phase II,SUCESS,10/05/2022,23/10/2022,1221,NaN,Kisumu
3,T1003,Paracetamol,Johnson & Johnson,Phase I,SUCESS,19/11/2021,2021-11-23,3312,293.29,Nairobi
4,T1004,Aspirin,Roche,Phase I,NaN,2021-11-11,18/04/2022,295,255.30,Nairobi


In [6]:
#now that i have read the pharma_data, i need to push it to Data folder while its still dirty 'Original form'

df.to_excel(r'C:\data_science\Python_programming_language\python_data_cleaning_project\python_pharma_data-cleaning\Data\pharma_dirty_dataset.xlsx',
index = False)

print(f'Saved {len(df)} rows to Data/pharma_dirty_dataset.xlsx') 

#just notised a warning 
#related to the "Data\pharma_dirty_dataset.xlsx" to fix it i will change '\' into forward slash '/'


df.head()

Saved 1200 rows to Data/pharma_dirty_dataset.xlsx


,Trial ID,Drug Name,Company,Trial Phase,Result,Start Date,End Date,Participants,Dosage (mg),Location
0,T1000,Metformin,Pfizer,Phase II,Ongoing,30/09/2022,2023-04-13,4084,376.55,Nairobi
1,T1001,Paracetamol,Johnson & Johnson,Phase III,SUCESS,01/02/2024,13/07/2024,3301,414.65,Kisumu
2,T1002,Aspirin,Novartis,Phase II,SUCESS,10/05/2022,23/10/2022,1221,NaN,Kisumu
3,T1003,Paracetamol,Johnson & Johnson,Phase I,SUCESS,19/11/2021,2021-11-23,3312,293.29,Nairobi
4,T1004,Aspirin,Roche,Phase I,NaN,2021-11-11,18/04/2022,295,255.30,Nairobi


**Initial inspection**

In [8]:
pharma_df = df  #renaming my df for easy understanding and clarity

pharma_df.head()

,Trial ID,Drug Name,Company,Trial Phase,Result,Start Date,End Date,Participants,Dosage (mg),Location
0,T1000,Metformin,Pfizer,Phase II,Ongoing,30/09/2022,2023-04-13,4084,376.55,Nairobi
1,T1001,Paracetamol,Johnson & Johnson,Phase III,SUCESS,01/02/2024,13/07/2024,3301,414.65,Kisumu
2,T1002,Aspirin,Novartis,Phase II,SUCESS,10/05/2022,23/10/2022,1221,NaN,Kisumu
3,T1003,Paracetamol,Johnson & Johnson,Phase I,SUCESS,19/11/2021,2021-11-23,3312,293.29,Nairobi
4,T1004,Aspirin,Roche,Phase I,NaN,2021-11-11,18/04/2022,295,255.30,Nairobi


In [ ]:
#inspecting pharma_df size/dimension

pharma_df.shape

#it has 1200 rows and 10 columns

(1200, 10)

In [ ]:
#inspecting the existing column names

pharma_df.columns


Index(['Trial ID', 'Drug Name', 'Company', 'Trial Phase', 'Result',
       'Start Date', 'End Date', 'Participants', 'Dosage (mg)', 'Location'],
      dtype='str')

 From the column names above we notice:

        1. names are in title case which we will change to snake case i prefer it better for column names
        2. double names are missing underscore, introducing underscore will reduce code break
        3. all the dtype = str yet we have some that are int and date type
        4. we can rename start and end dates to include trial on them. That way the dates will be more specific
        5. will need to fix dosage (mg) to something like dosage_mg that is if all data is in mg otherwise will need to standardize

In [13]:
pharma_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Trial ID      1116 non-null   str    
 1   Drug Name     1127 non-null   str    
 2   Company       1131 non-null   str    
 3   Trial Phase   1116 non-null   str    
 4   Result        1116 non-null   str    
 5   Start Date    1116 non-null   str    
 6   End Date      1117 non-null   str    
 7   Participants  1116 non-null   object 
 8   Dosage (mg)   1116 non-null   float64
 9   Location      1134 non-null   str    
dtypes: float64(1), object(1), str(8)
memory usage: 93.9+ KB


Additional Points to note:

    > There are missing values, we expect 1200 rows
    > dtype mixup will need to be fixed
    

**Data Cleaning steps**

        Step 1. Standardizing column names 

In [45]:
pharma_df.head()

,Trial ID,Drug Name,Company,Trial Phase,Result,Start Date,End Date,Participants,Dosage (mg),Location
0,T1000,Metformin,Pfizer,Phase II,Ongoing,30/09/2022,2023-04-13,4084,376.55,Nairobi
1,T1001,Paracetamol,Johnson & Johnson,Phase III,SUCESS,01/02/2024,13/07/2024,3301,414.65,Kisumu
2,T1002,Aspirin,Novartis,Phase II,SUCESS,10/05/2022,23/10/2022,1221,NaN,Kisumu
3,T1003,Paracetamol,Johnson & Johnson,Phase I,SUCESS,19/11/2021,2021-11-23,3312,293.29,Nairobi
4,T1004,Aspirin,Roche,Phase I,NaN,2021-11-11,18/04/2022,295,255.30,Nairobi


In [46]:
pharma_df_clean = pharma_df.clean_names(case_type = 'snake', strip_underscores = True)

#converted all column names into snake case and introduced underscores to link double names

pharma_df_clean.columns

Index(['trial_id', 'drug_name', 'company', 'trial_phase', 'result',
       'start_date', 'end_date', 'participants', 'dosage_mg', 'location'],
      dtype='str')

1.b renaming columns to make them relatable and scalable

In [47]:
pharma_df_clean.rename(
    columns = {
        'trial_id' : 'drug_trial_id',
        'company' : 'drug_manufacturer',
        'result' : 'trial_results',
        'start_date' : 'trial_start_date',
        'end_date' : 'trial_end_date',
        'participants' : 'trial_sample_size',
        'location' : 'trial_location'
    },
    inplace = True
)

#The idea is to make column names more relatable to non-technical team and also for sclability
#That is if this dataset was to be exapanded to 1000 column would the existing column make sense or would there exist
#conflicting names, For example location, is it the lab location, trial particapant location, customer location 


pharma_df_clean.columns

Index(['drug_trial_id', 'drug_name', 'drug_manufacturer', 'trial_phase',
       'trial_results', 'trial_start_date', 'trial_end_date',
       'trial_sample_size', 'dosage_mg', 'trial_location'],
      dtype='str')

Step 2 Correcting Data types

    a. change str to numerics for trial_sample_size
    

In [ ]:
#converting trial_sample_size from string to numeric

pharma_df_clean['trial_sample_size'] = pd.to_numeric(pharma_df_clean['trial_sample_size'], errors = 'coerce')

pharma_df_clean['trial_sample_size'].dtype

dtype('float64')

b. change dates from str to date time format

In [55]:
#pharma_df_clean[['trial_start_date', 'trial_end_date']]

pharma_df_clean['trial_start_date'].unique()

<StringArray>
['30/09/2022', '01/02/2024', '10/05/2022', '19/11/2021', '2021-11-11',
 '18/10/2021', '2022-01-12', '01/12/2021', '10/08/2022', '2020-12-22',
 ...
 '2020-02-21', '2022-08-14', '18/05/2022', '2020-08-05', '21/06/2021',
 '22/07/2022', '23/09/2020', '2020-07-06', '18/06/2020', '2021-10-14']
Length: 929, dtype: str

In [ ]:
pharma_df_clean['trial_end_date'].unique()

#we can see the date format here are mixed that is yyyy-mm-dd, dd-mm-yyyy, we have both '/' and '-'
#and its str note date formart

<StringArray>
['2023-04-13', '13/07/2024', '23/10/2022', '2021-11-23', '18/04/2022',
 '01/03/2022', '2022-06-02', '2022-03-05', '05/10/2022', '21/01/2021',
 ...
 '2020-07-20', '2023-03-29', '10/12/2022', '2021-10-02', '13/11/2020',
 '2020-06-08', '2021-09-25', '24/08/2020', '2021-03-15', '18/11/2020']
Length: 916, dtype: str

In [58]:
#1. we will fix the date separation character using regex to starndardize it

pharma_df_clean[['trial_start_date', 'trial_end_date']] = pharma_df_clean[['trial_start_date', 'trial_end_date']].replace('/', '-', regex = True)

pharma_df_clean[['trial_start_date', 'trial_end_date']]

,trial_start_date,trial_end_date
0,30-09-2022,2023-04-13
1,01-02-2024,13-07-2024
2,10-05-2022,23-10-2022
3,19-11-2021,2021-11-23
4,2021-11-11,18-04-2022
...,...,...
1195,23-09-2020,02-12-2020
1196,2020-07-06,24-08-2020
1197,09-11-2020,2021-03-15
1198,18-06-2020,18-11-2020


In [ ]:
#2. we will convert dates from str to datetime

for col in ['trial_start_date', 'trial_end_date']:
    pharma_df_clean[col] = pd.to_datetime(
        pharma_df_clean[col],
        errors = 'coerce',
        format = 'mixed',
        dayfirst = True
    )


pharma_df_clean[['trial_start_date', 'trial_end_date']] #looking at our output we can we see 

#we have cleaned our date columns

,trial_start_date,trial_end_date
0,2022-09-30,2023-04-13
1,2024-02-01,2024-07-13
2,2022-05-10,2022-10-23
3,2021-11-19,2021-11-23
4,2021-11-11,2022-04-18
...,...,...
1195,2020-09-23,2020-12-02
1196,2020-06-07,2020-08-24
1197,2020-11-09,2021-03-15
1198,2020-06-18,2020-11-18


In [ ]:
pharma_df_clean.info()

#Our columns are in correct data types formats

<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   drug_trial_id      1116 non-null   str           
 1   drug_name          1127 non-null   str           
 2   drug_manufacturer  1131 non-null   str           
 3   trial_phase        1116 non-null   str           
 4   trial_results      1116 non-null   str           
 5   trial_start_date   1116 non-null   datetime64[us]
 6   trial_end_date     1117 non-null   datetime64[us]
 7   trial_sample_size  1096 non-null   float64       
 8   dosage_mg          1116 non-null   float64       
 9   trial_location     1134 non-null   str           
dtypes: datetime64[us](2), float64(2), str(6)
memory usage: 93.9 KB


Step 3 Fix the case type of string columns

I have noticed some are using upper, others lower and others Title. I prefer title case for all of them

In [69]:
pharma_df_clean.head()

,drug_trial_id,drug_name,drug_manufacturer,trial_phase,trial_results,trial_start_date,trial_end_date,trial_sample_size,dosage_mg,trial_location
0,T1000,Metformin,Pfizer,Phase II,Ongoing,2022-09-30,2023-04-13,4084.0,376.55,Nairobi
1,T1001,Paracetamol,Johnson & Johnson,Phase III,SUCESS,2024-02-01,2024-07-13,3301.0,414.65,Kisumu
2,T1002,Aspirin,Novartis,Phase II,SUCESS,2022-05-10,2022-10-23,1221.0,NaN,Kisumu
3,T1003,Paracetamol,Johnson & Johnson,Phase I,SUCESS,2021-11-19,2021-11-23,3312.0,293.29,Nairobi
4,T1004,Aspirin,Roche,Phase I,NaN,2021-11-11,2022-04-18,295.0,255.30,Nairobi


In [70]:
pharma_df_clean['trial_phase'].unique()

<StringArray>
['Phase II', 'Phase III', 'Phase I', nan]
Length: 4, dtype: str

In [ ]:
pharma_df_clean['trial_results'].unique()

#we can see several issues here, mix of case types and several spelling errors 

<StringArray>
[  'Ongoing',   ' SUCESS',         nan,   'Failure',   'FAILURE',   'SUCCESS',
   'Success', 'SUCCCE SS',   'fAILURE']
Length: 9, dtype: str

In [72]:
#Here i will use a for loop to standardize case to title and trim leading and trailing spaces (whitespaces)

cols = ['drug_manufacturer', 'trial_results','trial_location']

for item in cols:
    pharma_df_clean[item] = pharma_df_clean[item].str.strip().str.title()


#since drug names follows scientific naming systems, leaving them in their original case is alwasy the best,
#the focus will therefore be on trimming the white spaces. 
#For the trial phase it incorporates roman numbers of which i would recommend leaving as they are in upper case
#only treaming whitespaces

for item in ['drug_name', 'trial_phase']: 
    pharma_df_clean[item] = pharma_df_clean[item].str.strip()

pharma_df_clean.head()

,drug_trial_id,drug_name,drug_manufacturer,trial_phase,trial_results,trial_start_date,trial_end_date,trial_sample_size,dosage_mg,trial_location
0,T1000,Metformin,Pfizer,Phase II,Ongoing,2022-09-30,2023-04-13,4084.0,376.55,Nairobi
1,T1001,Paracetamol,Johnson & Johnson,Phase III,Sucess,2024-02-01,2024-07-13,3301.0,414.65,Kisumu
2,T1002,Aspirin,Novartis,Phase II,Sucess,2022-05-10,2022-10-23,1221.0,NaN,Kisumu
3,T1003,Paracetamol,Johnson & Johnson,Phase I,Sucess,2021-11-19,2021-11-23,3312.0,293.29,Nairobi
4,T1004,Aspirin,Roche,Phase I,NaN,2021-11-11,2022-04-18,295.0,255.30,Nairobi


In [ ]:
pharma_df_clean['trial_results'].unique()

#after fixing the mixed cases and white spaces we can see there is still a typo! to fix this we will
#replace both 'Succce Ss', 'Sucess' with 'Success'

<StringArray>
['Ongoing', 'Sucess', nan, 'Failure', 'Success', 'Succce Ss']
Length: 6, dtype: str

In [75]:
pharma_df_clean['trial_results'] = pharma_df_clean['trial_results'].replace(['Sucess', 'Succce Ss'], 'Success')

pharma_df_clean['trial_results'].unique()

<StringArray>
['Ongoing', 'Success', nan, 'Failure']
Length: 4, dtype: str

In [76]:
#looking at the general view of the data at this point

pharma_df_clean.head()

,drug_trial_id,drug_name,drug_manufacturer,trial_phase,trial_results,trial_start_date,trial_end_date,trial_sample_size,dosage_mg,trial_location
0,T1000,Metformin,Pfizer,Phase II,Ongoing,2022-09-30,2023-04-13,4084.0,376.55,Nairobi
1,T1001,Paracetamol,Johnson & Johnson,Phase III,Success,2024-02-01,2024-07-13,3301.0,414.65,Kisumu
2,T1002,Aspirin,Novartis,Phase II,Success,2022-05-10,2022-10-23,1221.0,NaN,Kisumu
3,T1003,Paracetamol,Johnson & Johnson,Phase I,Success,2021-11-19,2021-11-23,3312.0,293.29,Nairobi
4,T1004,Aspirin,Roche,Phase I,NaN,2021-11-11,2022-04-18,295.0,255.30,Nairobi


**Step 4. Checking for duplicates and dropping them if they exist**

In [ ]:
#I will begin from the bigger picture. By scanning the dataset i see no duplicated columns
#secondly i will check if there is a duplicated full rows

pharma_df_clean.duplicated().sum #though this shows zero duplicate it could still be misleading because

#we are looking at all the columns

<bound method Series.sum of 0       False
1       False
2       False
3       False
4       False
        ...  
1195    False
1196    False
1197    False
1198    False
1199    False
Length: 1200, dtype: bool>

In [ ]:
#Now that we are sure their is no duplicate based on all the columns, we can go ahead
#and narrow down to a sizeable number of columns. This is mainly based on dorman expert advise
#or based on the context

#For example in our case, we can argue that a drug trial in the same phase with the same start date and end dates 
#could be a possible duplicate if it also has same dosage and was done in the same location


pharma_df_clean[
    pharma_df_clean[['drug_name', 'trial_phase', 'trial_start_date', 'trial_end_date',
'dosage_mg', 'trial_location']].duplicated()]

#This shows still there is no duplicates, if needed we can narrow down even further but only based on context


,drug_trial_id,drug_name,drug_manufacturer,trial_phase,trial_results,trial_start_date,trial_end_date,trial_sample_size,dosage_mg,trial_location


**Step 5. Checking for missing values and their imputation**

In [ ]:
pharma_df_clean.info()

#.info() gives a hint of existing missing values. we expect 1200 entries in each column but we have less than that across
#our dataset, that means missingness is spread across the entire dataset
# we can see there is a consistent pattern of 1116 entries, this is interesting and we need to investigate


<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   drug_trial_id      1116 non-null   str           
 1   drug_name          1127 non-null   str           
 2   drug_manufacturer  1131 non-null   str           
 3   trial_phase        1116 non-null   str           
 4   trial_results      1116 non-null   str           
 5   trial_start_date   1116 non-null   datetime64[us]
 6   trial_end_date     1117 non-null   datetime64[us]
 7   trial_sample_size  1096 non-null   float64       
 8   dosage_mg          1116 non-null   float64       
 9   trial_location     1134 non-null   str           
dtypes: datetime64[us](2), float64(2), str(6)
memory usage: 93.9 KB


In [ ]:
pharma_df_clean.isna().sum()

#upon further investigation, we can see trial_sample_size has the highest level of missingness, Thats 
#interesting because you can't carry out a trial without 'participants'

drug_trial_id         84
drug_name             73
drug_manufacturer     69
trial_phase           84
trial_results         84
trial_start_date      84
trial_end_date        83
trial_sample_size    104
dosage_mg             84
trial_location        66
dtype: int64

In [88]:
missingness_report = pd.DataFrame({
    'missing_count' : pharma_df_clean.isna().sum(),
    'missingness_percent' : round(pharma_df_clean.isna().mean() * 100 ,2)
})

missingness_report

,missing_count,missingness_percent
drug_trial_id,84,7.00
drug_name,73,6.08
drug_manufacturer,69,5.75
trial_phase,84,7.00
trial_results,84,7.00
trial_start_date,84,7.00
trial_end_date,83,6.92
trial_sample_size,104,8.67
dosage_mg,84,7.00
trial_location,66,5.50


In [92]:
pharma_df_clean['missing_per_row'] = pharma_df_clean.isna().sum(axis=1)

pharma_df_clean['missing_per_row'].value_counts()

missing_per_row
0    600
1    415
2    156
3     28
4      1
Name: count, dtype: int64

**Missing_per_row value counts interpretation**

- we can see half (600 rows) of our dataset is complete with no missing values.
- 415 rows are only missing one value (column)
- 156 rows are missing 2 values and 28 rows are missing 3 values which should be our concern
- 1 row has four missing values. This will need to be investigated further

From the above statistics we can see that the missingness is not random but has some pattern. The next step will be to narrow down and investgate the rows that have missing values

In [ ]:
pharma_df_clean.isna().corr() #this correlation, I want to understand if there is a clear pattern of missingnes

#That is if for example trial_phase is missing then trial_results will be missing too.
#but from the values. Majority seems to be very close to 0 that is (0.011645, -0.011265, 0.027138) 
# that is missing values have no relationship

#Values of 1 or close to 1 that is (0.99, 0.98) would show that missing value in one column would translate to 
#missing values in its correlated column

#values of -1 or close to -1 (-0.99, -0.98) 
# would mean missing values in one column influences the values to be present in onother column

,drug_trial_id,drug_name,drug_manufacturer,trial_phase,trial_results,trial_start_date,trial_end_date,trial_sample_size,dosage_mg,trial_location,missing_per_row
drug_trial_id,1.000000,-0.028832,-0.011645,-0.011265,0.027138,0.027138,0.015318,0.054794,-0.024066,0.048423,NaN
drug_name,-0.028832,1.000000,-0.017934,-0.001503,0.066818,0.012161,-0.028156,-0.004048,-0.042496,-0.046107,NaN
drug_manufacturer,-0.011645,-0.017934,1.000000,-0.011645,-0.011645,0.002385,0.017317,0.051148,-0.025675,-0.028185,NaN
trial_phase,-0.011265,-0.001503,-0.011645,1.000000,0.027138,0.001536,-0.010426,-0.014859,-0.011265,-0.008882,NaN
trial_results,0.027138,0.066818,-0.011645,0.027138,1.000000,-0.011265,-0.036170,-0.049686,0.014337,0.005444,NaN
trial_start_date,0.027138,0.012161,0.002385,0.001536,-0.011265,1.000000,0.015318,0.008358,0.001536,-0.037535,NaN
trial_end_date,0.015318,-0.028156,0.017317,-0.010426,-0.036170,0.015318,1.000000,-0.013930,0.028190,-0.036951,NaN
trial_sample_size,0.054794,-0.004048,0.051148,-0.014859,-0.049686,0.008358,-0.013930,1.000000,0.043185,-0.009354,NaN
dosage_mg,-0.024066,-0.042496,-0.025675,-0.011265,0.014337,0.001536,0.028190,0.043185,1.000000,0.005444,NaN
trial_location,0.048423,-0.046107,-0.028185,-0.008882,0.005444,-0.037535,-0.036951,-0.009354,0.005444,1.000000,NaN


In [ ]:
pharma_df_clean[pharma_df_clean['missing_per_row'] >= 3] #This affirms our correlation above

#That even when a row is missing 3 values or more its still not correlated

,drug_trial_id,drug_name,drug_manufacturer,trial_phase,trial_results,trial_start_date,trial_end_date,trial_sample_size,dosage_mg,trial_location,missing_per_row
64,NaN,Paracetamol,Johnson & Johnson,Phase I,Ongoing,NaT,2022-04-20,NaN,57.22,Eldoret,3
74,T1074,Ibuprofen,NaN,Phase I,Ongoing,2020-07-21,2020-08-11,NaN,75.37,NaN,3
83,T1083,Ibuprofen,NaN,Phase I,Success,2022-09-28,2022-11-10,NaN,409.24,NaN,3
194,T1194,Ibuprofen,Johnson & Johnson,Phase I,NaN,NaT,2024-06-14,3646.0,NaN,Nairobi,3
248,T1248,Ibuprofen,Pfizer,NaN,NaN,NaT,2021-02-14,1768.0,387.77,Mombasa,3
298,T1298,Paracetamol,Novartis,Phase II,Failure,2021-03-02,NaT,NaN,NaN,Kisumu,3
306,T1306,NaN,Pfizer,Phase II,Failure,NaT,2020-07-24,NaN,430.28,Eldoret,3
431,NaN,Aspirin,Novartis,Phase II,Success,2022-12-03,NaT,4000.0,NaN,Nairobi,3
458,NaN,NaN,Johnson & Johnson,Phase II,NaN,2020-08-15,2020-09-27,2845.0,411.58,Nairobi,3
528,NaN,Paracetamol,NaN,Phase II,Ongoing,2022-09-25,2023-12-04,NaN,484.58,Nairobi,3


**Missingness Report**

After carefull analysis of our missingness, we discover that though their was some level of consistent in count of missing values, the pattern is not clear. We also conclude that our missingness is not at random. At this point the best way forward is to consult dormain experts who entered the data to give us some more background. An explaination for this missingness

**columns to leave as they are**

1. trial_phase, if trial was not registered then its possible it wasn't assigned phase too.
2. trial_results, missing because results was not published. This also very critical and we would not want to extrapolate through imputation espcially because we are dealing with medicine. Facts are critical in this field.
3. Tria_start_date, maybe its missing because trial is not started yet.
4. trial_end_date, could be missing because trial is still ongoing or not started.
5. dosage_mg, missing because dose is not yet determined especially if trial is ongoing.
6. trial_sample_size, missing because in the early stages of trial, sample size is usually open that is not determined yet.

**columns to be imputed**

For the following columns we can impute missing values with 'Unknown' or 'Unspecified'

1. drug_name
2. drug_manufacturer
3. trial_location
4. drug_trial_id


In [ ]:
pharma_df_clean['drug_trial_id'].unique()

#this uncovers that our drug_trial_id is a continous identifier (serial) in that regard we will be able to impute
#its missing values without distorting the dataset

<StringArray>
['T1000', 'T1001', 'T1002', 'T1003', 'T1004', 'T1005', 'T1006', 'T1007',
 'T1008', 'T1009',
 ...
 'T2188', 'T2189', 'T2191', 'T2192', 'T2193', 'T2195', 'T2196', 'T2197',
 'T2198', 'T2199']
Length: 1113, dtype: str

**Fixing the drug_trial_id missingness**

step 1. Extract numeric part

In [99]:
pharma_df_clean['id_num'] = pharma_df_clean['drug_trial_id'].str.extract(r'(\d+)').astype(float)

pharma_df_clean['id_num'].head()

0    1000.0
1    1001.0
2    1002.0
3    1003.0
4    1004.0
Name: id_num, dtype: float64

Step 2. interpolate missing values

In [100]:
pharma_df_clean['id_num'] = pharma_df_clean['id_num'].interpolate().astype(int)

pharma_df_clean['id_num'].info()

<class 'pandas.Series'>
RangeIndex: 1200 entries, 0 to 1199
Series name: id_num
Non-Null Count  Dtype
--------------  -----
1200 non-null   int64
dtypes: int64(1)
memory usage: 9.5 KB


Step 3. Add 'T' back to rebuild the trial ID

In [102]:
pharma_df_clean['drug_trial_id'] = 'T' + pharma_df_clean['id_num'].astype(str)

pharma_df_clean['drug_trial_id'].isna().sum()

np.int64(0)

Step 4. Drop the id_num now that we no longer need it


In [103]:
pharma_df_clean = pharma_df_clean.drop(columns = ['id_num'])

pharma_df_clean.head()

,drug_trial_id,drug_name,drug_manufacturer,trial_phase,trial_results,trial_start_date,trial_end_date,trial_sample_size,dosage_mg,trial_location,missing_per_row
0,T1000,Metformin,Pfizer,Phase II,Ongoing,2022-09-30,2023-04-13,4084.0,376.55,Nairobi,0
1,T1001,Paracetamol,Johnson & Johnson,Phase III,Success,2024-02-01,2024-07-13,3301.0,414.65,Kisumu,0
2,T1002,Aspirin,Novartis,Phase II,Success,2022-05-10,2022-10-23,1221.0,NaN,Kisumu,1
3,T1003,Paracetamol,Johnson & Johnson,Phase I,Success,2021-11-19,2021-11-23,3312.0,293.29,Nairobi,0
4,T1004,Aspirin,Roche,Phase I,NaN,2021-11-11,2022-04-18,295.0,255.30,Nairobi,1


**Fixing missing values in the following columns**

1. drug_name
2. drug_manufacturer
3. trial_location

In [104]:
cols = ['drug_name', 'drug_manufacturer', 'trial_location']

for item in cols:
    pharma_df_clean[item] = pharma_df_clean[item].fillna('Unknown')


pharma_df_clean[['drug_name', 'drug_manufacturer', 'trial_location']].isna().sum()

drug_name            0
drug_manufacturer    0
trial_location       0
dtype: int64

**Fixing missing values for drug_sample_size and dosage_mg**

While i did recommend leaving these columns as they are, since this is a very sensitive dataset that involves medicines for humans. An alternative creative imputation to facilitate research is neccesary.

My recommendation therefore, will be to find the mean of these columns based on specific groupby columns:

-  for drug_sample_size. to find the right mean to impute we will need to group by (drug_name, drug_manufacturer, and trial_phase) this will give us the closest value that we are missing. 

    - Trials of the same drug
    - From the same manufacturer
    - In the same phase

                 **Potential challenges with this plan**

If a group has only 1 or 2 rows, the “mean” is meaningless.

Example: Drug A, Manufacturer X, Phase II → only 1 trial
Missing sample size → mean = that one value
Missing dosage → mean = that one value
This is will not imputation but duplication.

- for dosage_mg. I recommend we use group by (drug_name, drug_manufacturer, trial_phase, and trial_sample_size). The resulting mean will be very specific and the closest value we can get to impute our missing value